src/LegacyThumbnailMigrationHelper.php

```diff
@@ -0,0 +1,197 @@
+        $folders = $this->getFolderQuery();
+        foreach ($folders->dataQuery()->execute() as $folderData) {
+            // More memory efficient way to looping through large sets
+            /** @var Folder $folder */
+            $folder = $folders->createDataObject($folderData);
+            $moved = $this->migrateFolder($store, $folder);
+            if ($moved) {
+                $allMoved = array_merge($allMoved, $moved);
+            }
+        }
+        if (class_exists(Versioned::class)) {
+            Versioned::set_reading_mode($originalState);
+        }
+        return $allMoved;
+    }
+
+    /**
+     * @param LoggerInterface $logger
+     */
+    public function setLogger(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
+
+        return $this;
+    }
+
+    /**
+     * Migrate a folder
+     *
+     * @param AssetStore $store
+     * @param Folder $folder
+     * @param string $legacyFilename
+     * @return array Map of old to new file paths (relative to store root)
+     */
+    protected function migrateFolder(AssetStore $store, Folder $folder)
+    {
+        $moved = [];
+
+        // Get normalised store relative path
+        $folderPath = preg_replace('#^/#', '', $folder->getFilename());
+        $resampledFolderPath = $folderPath . '_resampled';
+
+        // Legacy thumbnails couldn't have been stored in a protected filesystem
+        /** @var \League\Flysystem\Filesystem $filesystem */
+        $filesystem = $store->getPublicFilesystem();
+
+        /** @var FileResolutionStrategy $strategy */
+        $strategy = $store->getPublicResolutionStrategy();
+
+        if (!$filesystem->has($resampledFolderPath)) {
+            return $moved;
+        }
+
+        // Recurse through folder
+        foreach ($filesystem->listContents($resampledFolderPath, true) as $fileInfo) {
+            if ($fileInfo['type'] !== 'file') {
+                continue;
+            }
+
+            $oldResampledPath = $fileInfo['path'];
+
+            // Get "variant folders" inside _resampled folder.
+            $variantFolders = preg_replace('#.*_resampled/(.*)#', '$1', $fileInfo['dirname']);
+            $encodedVariants = explode(DIRECTORY_SEPARATOR, $variantFolders);
+
+            // Replicate new variant format.
+            // We're always placing these files in the public filesystem, *without* a content hash path.
+            // This means you need to run the optional migration task introduced in 4.4,
+            // which moves public files out of content hash folders.
+            // Image->manipulate() is in charge of creating the full file name (incl. variant),
+            // and assumes the manipulation is run on an existing original file, so we can't use it here.
+            // Any AssetStore-level filesystem operations (like move()) suffer from the same limitation,
+            // so we need to drop down to path based filesystem renames.
+            $newResampledPath = implode('', [
```

Comment:
No, the point was that it shouldn't matter if the new file exists. It might have been deleted in the meantime. Due to the shortcodes issues (https://github.com/silverstripe/silverstripe-assets/issues/167), existing thumbnails have lost their connection to the original file. So its entirely possible that thumbnails are still relied on in e.g. HTML content, but no longer available through the original file. Technically those are now orphaned files with no way to clean them up through SS APIs, but that's better than breaking images on website content.